### PR TITLE
Move engines to Warp

### PR DESCRIPTION
**Pull request by gakonst**
As title, same principle as https://github.com/interledger-rs/interledger-rs/pull/420

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -1,6 +1,11 @@
+/// # Settlement Engine API
+///
+/// Tower_Web service which exposes settlement related endpoints as described in RFC536,
```
Comment (emschwartz) [inline src/api.rs]: This comment should be updated

File: src/api.rs [hunk 2]
```diff
@@ -9,145 +14,235 @@ use hyper::{Response, StatusCode};
...
-            let input = format!("{:?}", body);
-            let input_hash = get_hash_of(input.as_ref());
-            let engine = self.engine.clone();
-            let f = move || engine.create_account(body);
-            self.make_idempotent_call(f, input_hash, idempotency_key)
-        }
+// Helper function that returns any idempotent data that corresponds to a
+// provided idempotency key. It fails if the hash of the input that
+// generated the idempotent data does not match the hash of the provided input.
+fn check_idempotency<S>(
```
Comment (emschwartz) [inline src/api.rs]: If we're going to make this crate depend on `interledger-api`, what do you think about having that crate export shared functionality like this (also behind a feature flag)?

**Review by emschwartz — APPROVED** [review]
LGTM. @dora-gt should review as well

File: src/api.rs (hunk 2 above)
Comment (gakonst) [reply src/api.rs]: I agree with that.

File: Cargo.toml [hunk 3]
```diff
@@ -7,20 +7,20 @@ license = "Apache-2.0"
...
-tower-web = "0.3.7"
 hex = "0.3.2"
 web3 = "0.8.0"
 log = "0.4.6"
 tokio = "0.1.21"
 hyper = "0.12.31"
 futures = "0.1.25"
 interledger-service = "^0.2.2-alpha.1"
-interledger-settlement = { git = "https://github.com/interledger-rs/interledger-rs", branch = "master" }
+interledger-settlement = { git = "https://github.com/interledger-rs/interledger-rs", branch = "gakonst/settlement-warp" }
```
Comment (emschwartz) [inline Cargo.toml:17]: This does make me think that `interledger-settlement` should be included in this repo

**Review by emschwartz — COMMENTED** [review]
Gah I hate when I review something and then it doesn't get submitted and stays pending. Sorry for the delay

File: src/api.rs [hunk 4]
```diff
@@ -1,194 +1,174 @@
...
+    E: SettlementEngine + Clone + Send + Sync + 'static,
+    S: IdempotentStore + Clone + Send + Sync + 'static,
+{
+    let with_store = warp::any().map(move || store.clone()).boxed();
+    let with_engine = warp::any().map(move || engine.clone()).boxed();
+    let idempotency = warp::header::optional::<String>("idempotency-key");
+    let account_id = warp::path("accounts").and(warp::path::param2::<String>()); // account_id
+
+    // POST /accounts/ (optional idempotency-key header)
+    // Body is a Vec<u8> object
```
Comment (dora-gt) [inline src/api.rs:39]: Is this comment right? It seems `.and(warp::body::json())`.

File: src/api.rs [hunk 5]
```diff
@@ -284,68 +262,37 @@ mod tests {
     fn idempotent_execute_settlement() {
         let store = test_store();
         let engine = TestEngine;
-        let api = SettlementEngineApi {
-            store: store.clone(),
-            engine,
+        let api = create_settlement_engine_filter(engine, store.clone());
+
+        let settlement_call = move |id, amount, scale| {
+            warp::test::request()
```
Comment (dora-gt) [inline src/api.rs:268]: 👀

I'll consider doing this next time with `TestStore`.

**Review by dora-gt — APPROVED** [review]
LGTM, just left a comment.

File: Cargo.toml (hunk 3 above)
Comment (gakonst) [reply Cargo.toml:17]: Agreed.

**Pull request merged**